### ISSUE TITLE
Setup testing for scratch-gui repo

### ISSUE THREAD

**Issue by bogusred** [LLK/scratch-gui#421]
My understanding is that the scratch-gui repo doesn't have any test setup currently. I'd like to help by drafting a design doc (unless you have one already?) and help do some research. If the scratch team reaches a decision on type of testing, testing framework, etc I can help get the implementation started.

If there has been any research/notes on this already, please let me know.

[Design doc draft](https://docs.google.com/document/d/1Q77jROqZyOaVjHicwNlmgyE_OIlE7rBlczzj6miok3w/edit)

From Ray:

> Unit would be most useful at the beginning. There are several frameworks for doing that for React, but we haven't investigated the ins and outs of each, which is something that would be really helpful to do on that issue.
> 
> And for questions about scratch-gui me and paul would be the best people to ask

**Comment by bogusred** [comment]
## Status update:
### Things completed today:

- Draft of [design doc](https://docs.google.com/document/d/1Q77jROqZyOaVjHicwNlmgyE_OIlE7rBlczzj6miok3w/edit#) shared with Ray and Paul from scratch team
- Initial research on various testing libraries and setups: jasmine+ karma, mocha, jest, enzyme. Findings from research are in the design doc.
- Experimentation with setting up Mocha and Enzyme. Here's [a repo with what i have so far](https://github.com/bogusred/scratch-gui/tree/feature/setup-testing) and [my notes](https://docs.google.com/document/d/1Q77jROqZyOaVjHicwNlmgyE_OIlE7rBlczzj6miok3w/edit#heading=h.aamrd2ohvdbo).
- Initial proposal on direction. TL;DR:  Seems like the most robust and popular option would be Mocha and Enzyme (designed specifically for testing react). 


### Blockers

- need decision from Scratch team on requirements for testing. 
- need decision from Scratch team on decision for testing library direction (will come later after defining requirements)
- blocked on experimentation of mocha setup.  I don't really have experience with configuring webpack (since I usually use the Angular cli which does this for me) and I suspect this may be the reason I've run into issues. If someone more experienced wants to take a look that would help.

Also as a note, I have no react development experience. I also have no experience setting up Mocha but I'm hoping my experience setting up Jasmine and Karma will help.

**Comment by bogusred** [comment]
Just a heads up I probably wont have a significant update until the weekend when i'll have more time to continue experimenting. in the meantime i'll try to reach out to any contacts i have that have react experience and get their advice/input.

**Comment by paulkaplan** [comment]
Thanks again for starting this! I added a few comments, an example test I wrote and a request to recreate some example tests using different frameworks.

**Comment by bogusred** [comment]
Thank you Paul.
I spent some time experimenting today with both tap and jest.

Tap I ran into roadblocks. It can't load the component .jsx files (it throws a syntax error). Presumably because there needs to be some sort of pre-compile step? If you have any suggestions on how to achieve this let me know. Also Tap is incredibly limiting out of the box. we'd need to find additional third party libs to get this working (for rendering react components, triggering clicks, spies).

I got a hello world working with Jest. I'll try to get further with the Jest setup tomorrow. Unfortunately I lost a lot of time trying to debug windows issues. Jest isn't running for me in the scratch-gui repo when using git bash. But I was able to get it working in windows command line. (updating node didn't help). Details are in the design doc.
